src/poly1305.rs

```diff
@@ -87,27 +142,69 @@ pub type Tag = [u8; TAG_LEN];
 /// The length of a `Tag`.
 pub const TAG_LEN: usize = 128 / 8;
 
-type State = [u8; STATE_LEN];
-const STATE_LEN: usize = 256;
+const BLOCK_STATE_SIZE: usize = 192;
+
+#[repr(C)]
+struct Funcs {
+    blocks_fn: unsafe extern fn(*mut c_void, *const u8, c::size_t, u32), 
+    emit_fn: unsafe extern fn(*mut c_void, *mut u8, *const u32),
+}
+
+fn init(state: &mut [u8; BLOCK_STATE_SIZE], key: &[u8; KEY_LEN], func: *mut Funcs) -> i32 {
+    unsafe {
+        GFp_poly1305_init_asm(
+            state.as_mut_ptr() as *mut c_void,
+            key.as_ptr(),
+            func as *mut c_void
+        )
+    }
+}
+
+impl Funcs {
+    fn blocks(&self, state: &mut [u8; BLOCK_STATE_SIZE], data: &[u8], should_pad: u32) {
+        unsafe {
+            (self.blocks_fn)(
+                state.as_mut_ptr() as *mut c_void,
+                data.as_ptr(),
+                data.len(),
+                should_pad
+            )
+        };
+    }
+
+    fn emit(&self, state: &mut [u8; BLOCK_STATE_SIZE], tag: &mut Tag, nonce: &[u32; 4]) {
+        unsafe {
+             (self.emit_fn)(
+                 state.as_mut_ptr() as *mut c_void,
+                 tag.as_mut_ptr(),
+                 nonce.as_ptr()
+             );
+        }
+    }
+}
+
+#[repr(C)]
+pub struct SigningContext {
+    _align: [u64; 0],
+    opaque: [u8; BLOCK_STATE_SIZE],
+    nonce: [u32; 4],
+    buf: [u8; 16],
+    buf_used: c::int,
+    func: Funcs
+}
 
 extern {
-    fn GFp_poly1305_init(state: &mut State, key: &KeyBytes);
-    fn GFp_poly1305_finish(state: &mut State, mac: &mut Tag);
-    fn GFp_poly1305_update(state: &State, in_: *const u8, in_len: c::size_t);
+    fn GFp_poly1305_init_asm(state: *mut c_void, key: *const u8, out_func: *mut c_void) -> c::int;
```

Comment:
if `state` needs to be 64-bit aligned then we should note it by defining a type named `State` as an alias for `[u64; (BLOCK_STATE_SIZE + 7) / 8]` with a comment explaining it needs to be aligned.

Also, does the key need to be 64-bit aligned?

I believe you can write this as:
```rust
    fn GFp_poly1305_init_asm(state: &mut State, key: &KeyBytes, out_func: &mut Func) -> c::int;
```

Please document what the return type signifies.

And similar for the other functions.